juju: Add status for deprecated kubelet args

Cynerva opened the pull request:
This adds a `blocked` status to inform the user when `kubelet-extra-args` has been configured with an arg that is deprecated in kubelet. Same for `proxy-extra-args` and kube-proxy.

For example:
```
juju config kubernetes-worker kubelet-extra-args='anonymous-auth=false`
```

results in the following charm status:
```
kubelet-extra-args: anonymous-auth is deprecated
```

hyperbolic2346 commented on cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py:
```diff
@@ -347,10 +349,51 @@ def charm_status(kube_control):
+        extra_args = parse_extra_args(config_key)
+        for arg in extra_args:
+            if arg in arg_help and 'DEPRECATED:' in arg_help[arg]:
+                deprecated_args.append((config_key, arg))
```
Should we call out invalid args here? Seems if arg is not in arg_help there is a problem.

Cynerva replied:
Not necessarily. Kubelet 1.8.x had a hidden `experimental-dockershim` flag that didn't show in help output, for example.

I don't see any hidden args in 1.9.x, 1.10.x or 1.11.x, but I don't want to assume that will always be the case.

Might be worthwhile to log a warning though. What do you think?

johnsca commented on cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py:
```diff
@@ -347,10 +349,51 @@ def charm_status(kube_control):
+        # this isn't vital, log it and move on
+        traceback.print_exc()
+    if deprecated_args:
+        msg = '%s: %s is deprecated' % deprecated_args[0]
```
Could also do `'deprecated args: %s' % ', '.join(deprecated_args)` but that could get pretty long.  I'm +1 on the number & check logs message.

johnsca commented on cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py:
```diff
@@ -360,12 +403,12 @@ def update_kubelet_status():
-    if len(failing_services) == 0:
-        hookenv.status_set('active', 'Kubernetes worker running.')
-    else:
+    if failing_services:
```
Out of scope for this PR, but since you touched this bit too, something I hit yesterday was the charm reporting status OK when none of the system pods were running.  Might be worth thinking about checking for those as well, or some other minimal health check.

Cynerva commented:
Pushed changes based on feedback:

* When an arg is missing from help output, log a warning (but don't show a status message since it could be a legit configuration)
* For each deprecated arg, log a warning, e.g. `kubelet-extra-args: anonymous-auth is deprecated`
* Change the status message to the following format: `kubelet-extra-args: anonymous-auth is deprecated (+3 others, see juju debug-log)`

@hyperbolic2346 @johnsca please let me know if this addresses your concerns. Thanks!

hyperbolic2346 commented:
LGTM

Cynerva commented:
I've rebased this to clean up the merge commits. Should be ready to go once again.
